sig-api-machinery/extensible-admission-to-beta.md

```diff
@@ -0,0 +1,362 @@
+# Admission control changes
+
+Admission control is the key security and cluster policy enforcement mechanism
+in Kubernetes. In order to make a reliable extension mechanism for admission,
+we must reconsider the current chain to produce understandable extensions.
+
+We are proposing a series of changes to admission control in an attempt to make
+it less error-prone, easier to configure, and more flexible for admission plugin
+developers.
+
+## Requirements for extensible admission
+Ordering the admission chain is critical.  One mutating admission plugin can invalidate
+the results of other admission plugins.  For instance, the `LimitVerifier` would
+invalidate results for the `ResourceQuota` plugin.  More generally, object mutations
+must happen before object validations.  In addition, compiled validating admission plugins
+must run after mutating admission webhooks.  Splitting the chain into mutators and validators
+will resolve the coarse grained ordering problem and would require two different webhook APIs
+since the `Status` would be different for both now and we can reasonably expect different
+evolutions.  It would also force changes in the registration APIs since ordering matters
+for individual mutators, but ordering does not matter for validators.  Having fewer items to 
```

Comment:
> ordering does not matter for validators

it could matter for remote systems, if the webhook is non-mutating from kube's perspective of the object being admitted, but has side effects outside of kube (like quota)

Reply:
> it could matter for remote systems, if the webhook is non-mutating from kube's perspective of the object being admitted, but has side effects outside of kube (like quota)

Providing order on validators would have significant performance implications.  Even in your quota example, it's not important for quota to be ordered.  It's a performance optimization if something validates "no".  Quota (and things like it), still have to have reconciliation loops since failures could happen down in storage.

Reply:
> Providing order on validators would have significant performance implications

Not providing order could also have significant performance implications, if the external check has an expensive side effect.

> Even in your quota example, it's not important for quota to be ordered. It's a performance optimization if something validates "no".

Performance optimizations can be important.